runtime/js/test.js

```diff
@@ -0,0 +1,209 @@
+
+function getGlobalRoot() {
+  if (!globalRoot) {
+    globalRoot = {
+      started: Date.now(),
+      pendingTests: [],
+      result: {
+        passed: 0,
+        failed: [],
+      },
+      lastReportedTest: undefined,
+    };
+    setTimeout(runNextTest);
+  }
+  return globalRoot;
+}
+
+/** @typedef {{
+  fileName: string;
+  lineNumber: number;
+  columnNumber: number;
+}} TestLocation */
+
+class TestCase {
+  /**
+   * @param {Object} args
+   * @param {string} args.name
+   * @param {() => void | () => Promise<void>} args.fn
+   * @param {TestLocation} args.location
+   */
+  constructor({ name, fn, location }) {
+    this.name = name;
+    this.fn = fn;
+    this.location = location;
+    this.started = undefined;
+  }
+
+  async execute() {
+    this.started = Date.now();
+    let maybePromise;
+    try {
+      maybePromise = this.fn();
+    } catch (error) {
+      this.#recordFailure(error);
+      return;
+    }
+
+    if (typeof maybePromise?.then !== "function") {
+      // The test fn was synchronous, we are done.
+      this.#recordSuccess();
+      return;
+    }
+
+    // The test fn returned a promise, we need to wait for completion
+    return maybePromise.then(
+      () => this.#recordSuccess(),
+      (error) => this.#recordFailure(error),
+    );
+  }
+
+  #log(passed) {
+    if (globalRoot.lastReportedTest?.location.fileName !== this.location.fileName) {
+      // TODO(bajtos) Ideally, we should report paths relative to the project root
+      console.log("\n%s", this.location.fileName.split(/[\\\/]+/g).slice(-1)[0]);
+      globalRoot.lastReportedTest = this;
+    }
+
+    let duration = Date.now() - this.started;
+    console.log(
+      "  %s %s %s",
+      passed ? green("\u2714" /* check mark */) : red("\u2716" /* heavy multiplication */),
+      this.name,
+      grey(`(${duration}ms)`),
+    );
+  }
+
+  #recordFailure(error) {
+    globalRoot.result.failed.push(new TestFailure({ testCase: this, error }));
+    this.#log(false);
+  }
+
+  #recordSuccess() {
+    globalRoot.result.passed++;
+    this.#log(true);
+  }
+}
+
+/**
+ * @param {TestLocation} location
+ * @returns {string}
+ */
+function displayTestLocation({ fileName, lineNumber, columnNumber }) {
+  return `${fileName}:${lineNumber}:${columnNumber}`;
+}
+
+class TestFailure {
+  /**
+   * @param {Object} args
+   * @param {TestCase} args.testCase
+   * @param {Error} args.error
+   */
+  constructor({ testCase, error }) {
+    this.testCase = testCase;
+    this.error = error;
+  }
+
+  render() {
+    let location = grey(`=> ${displayTestLocation(this.testCase.location)}`);
+    return `${this.testCase.name} ${location}\n${boldRed("error:")} ${
+      this.error.stack || this.error
+    }`;
+  }
+}
+
+function runNextTest() {
+  const nextTest = globalRoot.pendingTests.shift();
+
+  if (nextTest) {
+    nextTest.execute(nextTest).then(runNextTest);
+    return;
+  }
+
+  reportTestResults();
+}
+
+function reportTestResults() {
```

Comment:
Discussion point.

There are no tests for this test runner and the test reporter. We have a certain amount of implicit code coverage thanks to running Zinnia tests using this same test runner. Is it enough, though?

Should I add a few tests to verify the most important scenarios, including tests checking the output printed to stdout & stderr?

Reply:
Output checking sgtm! This is also the primary way the node core test library is tested

Reply:
If we don't have that and only the "living spec", then it will be tricky to say if an error is because of a test bug or a runner bug

Reply:
Makes sense. Adding tests for the test runner's output required more work on our testing infrastructure, but it was worth it. See 7a29089.